### DISCUSSION TITLE
How to implement the term (stress multiply strain rate) in MOOSE?

### DISCUSSION THREAD

**Discussion by rh201** [idaholab/moose#18822] [Q&A General] [answered]
Hi MOOSEer,

I want to ask how to calculate the term below in MOOSE? Are there any similar examples on this? Thanks.

![image](https://user-images.githubusercontent.com/79764031/133285593-9a408a06-3fe4-422d-881e-b09fb2404c37.png)
where \sigma is stress and \epsilon is strain.

**Comment by GiudGiud** [comment]
They are both tensors right?
This documentation can help you here https://mooseframework.inl.gov/docs/doxygen/moose/classRankTwoTensorTempl.html

**Reply by rh201** [reply]
@GiudGiud Thanks! Yes, they are both ranktwo tensor, I'll have look into this page. Are there any examples of it?

**Reply by GiudGiud** [reply]
For sure. Once you find the operation you should be using (outer_product I think?) then you can grep the repo for it. I dont think there are examples in the tutorials.

**Reply by rh201** [reply]
OK, thank you:) I'll have a try and see what happens.

**Comment by ajacquey** [comment, marked as answer]
I think you are looking for the `double_contraction()` function.

It looks like you wants the [`Plastic Heat Energy`](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/kernels/PlasticHeatEnergy.C) kernel and associated [`ComputePlasticHeatEnergy`](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/ComputePlasticHeatEnergy.C) material from the tensor mechanics module.

If you'd prefer to implement your own, have a look at lines 42-43 from [`ComputePlasticHeatEnergy`](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/ComputePlasticHeatEnergy.C).

**Reply by rh201** [reply]
@ajacquey That's very similar to what I want, I'll definitely learn the pages. Thank you !

**Reply by rh201** [reply]
@ajacquey Hi! When I try to modify the material module above (ComputePlasticHeatEnergy), a variable **_d_stress_dT** (the derivative of stress w.r.t temperature) was defined, and the rest is mostly the same. But it has the error message as:

> error: 'const class MaterialProperty<RankTwoTensorTempl<double> >' has no member named 'doubleContraction'
   57 |       _d_stress_dT[_qp].doubleContraction(_elastic_strain[_qp] - _elastic_strain_old[_qp]) / _dt;

Do you have any idea of what's going on with the error? Thanks.

**Reply by ajacquey** [reply]
Hi @rh201,
Not sure what causes this error. 
Could you share your files?

**Reply by rh201** [reply]
This is the .C file modified form ComputePlasticHeatEnergy, in which I want to consider **_d_stress_dT** and AD version.


```
InputParameters
ComputeElasticHeatEnergy::validParams()
{
  InputParameters params = ADMaterial::validParams();
  params.addParam<std::string>("base_name",
                               "Optional parameter that allows the user to define "
                               "multiple mechanics material systems on the same "
                               "block, i.e. for multiple phases");
  params.addRequiredCoupledVar("temperature","Coupled temperature variable");
  params.addClassDescription("elastic heat energy density = stress * elastic_strain_rate");
  return params;
}

ComputeElasticHeatEnergy::ComputeElasticHeatEnergy(const InputParameters & parameters)
  : DerivativeMaterialInterface<ADMaterial>(parameters),
    // _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
    // _plastic_strain(getMaterialProperty<RankTwoTensor>("plastic_strain")),
    // _plastic_strain_old(getMaterialPropertyOld<RankTwoTensor>("plastic_strain")),
    // _stress(getMaterialProperty<RankTwoTensor>(_base_name + "stress")),
    // _Jacobian_mult(getMaterialProperty<RankFourTensor>(_base_name + "Jacobian_mult")),
    // _elasticity_tensor(getMaterialProperty<RankFourTensor>(_base_name + "elasticity_tensor")),
    // _plastic_heat(declareProperty<Real>(_base_name + "plastic_heat")),
    // _dplastic_heat_dstrain(declareProperty<RankTwoTensor>(_base_name + "dplastic_heat_dstrain"))
    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
    _elastic_strain(getADMaterialProperty<RankTwoTensor>("elastic_strain")),
    _elastic_strain_old(getMaterialPropertyOld<RankTwoTensor>("elastic_strain")),
    _stress(getADMaterialProperty<RankTwoTensor>(_base_name + "stress")),
    _d_stress_dT(parameters.hasDefaultCoupledValue("temperature")
          ? nullptr
          : &getADMaterialProperty<RankTwoTensor>(derivativePropertyNameFirst(_base_name + "stress",
             getVar("temperature", 0)->name()))),
    _T_dot(adCoupledDot("temperature")),
    // _Jacobian_mult(getADMaterialProperty<RankFourTensor>(_base_name + "Jacobian_mult")),
    // _elasticity_tensor(getADMaterialProperty<RankFourTensor>(_base_name + "elasticity_tensor")),
    _elastic_heat(declareADProperty<Real>(_base_name + "elastic_heat"))
    // _delastic_heat_dstrain(declareADProperty<RankTwoTensor>(_base_name + "delastic_heat_dstrain"))
{
}

void
ComputeElasticHeatEnergy::computeQpProperties()
{
  _elastic_heat[_qp] =
      _d_stress_dT[_qp].doubleContraction(_elastic_strain[_qp] - _elastic_strain_old[_qp]) / _dt;

  // if (_fe_problem.currentlyComputingJacobian())
  // {
  //   if (_elastic_strain[_qp] == _elastic_strain_old[_qp])
  //     // no plastic deformation, so _elasticity_tensor = _Jacobian_mult
  //     _delastic_heat_dstrain[_qp] = RankTwoTensor();
  //   else
  //   {
  //     _delastic_heat_dstrain[_qp] =
  //         (_elastic_strain[_qp] - _elastic_strain_old[_qp]).initialContraction(_Jacobian_mult[_qp]);
  //     _delastic_heat_dstrain[_qp] += _stress[_qp];
  //     _delastic_heat_dstrain[_qp] -=
  //         _stress[_qp].initialContraction(_elasticity_tensor[_qp].invSymm() * _Jacobian_mult[_qp]);
  //     _delastic_heat_dstrain[_qp] /= _dt;
  //   }
  // }
}
```
The error I mentioned above will show when compiling.



And when I call elastic_strain_old with **getADMaterialPropertyOld** as below, 
`    _elastic_strain_old(getADMaterialPropertyOld<RankTwoTensor>("elastic_strain")),`
another error will occur

> error: 'getADMaterialPropertyOld' was not declared in this scope; did you mean 'getMaterialPropertyOld'?
   39 |     _elastic_strain_old(getADMaterialPropertyOld<RankTwoTensor>("elastic_strain")),
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~
      |                         getMaterialPropertyOld

It seems there are no **getADMaterialPropertyOld** function.


Hope it's clear :)

**Reply by GiudGiud** [reply]
There is no getADMaterialPropertyOld, AD terms are not kept for the old time step. You need to retrieve this as a regular material property
See this https://github.com/idaholab/moose/discussions/18482

**Reply by hugary1995** [reply]
@rh201 , what @GiudGiud said is correct.

Your code doesn't compile because your `_d_stress_dT` is a pointer. You should instead use
```
(*_d_stress_dT)[_qp].doubleContraction(_elastic_strain[_qp] - _elastic_strain_old[_qp]) / _dt;
```

**Reply by rh201** [reply]
@GiudGiud @hugary1995 Thank you for pointing this out. It helps. 

BTW, how should I define and calculate **_d_stress_dT** inside the .C file instead of in input file? since I think it's not convenient to directly retrieve stress when using DerivativeParsedMaterial in input file.

**Reply by hugary1995** [reply]
Suppose your stress is a `MaterialProperty<RankTwoTensor>`, and your temperature is a `VariableValue`, you can declare the derivative of stress w.r.t. temperature (in the initialization list) using
```cpp
_dstress_dT(declareProperty<RankTwoTensor>(derivativePropertyName("stress", {getVar("temperature", 0)->name()})))
```
where the method `derivativePropertyName` is defined in `DerivativeMaterialPropertyNameInterface`.

Similarly, if you want to use AD, you can use
```cpp
_dstress_dT(declareADProperty<RankTwoTensor>(derivativePropertyName("stress", {getVar("temperature", 0)->name()})))
```

**Reply by rh201** [reply]
@hugary1995  Thanks for your help, it compiles successfully now :)
